source/common/router/router.cc

```diff
@@ -32,6 +32,58 @@ namespace Envoy {
 namespace Router {
 namespace {
 uint32_t getLength(const Buffer::Instance* instance) { return instance ? instance->length() : 0; }
+
+bool schemeIsHttp(const Http::HeaderMap& downstream_headers,
+                  const Network::Connection* connection) {
+  if (downstream_headers.ForwardedProto() && downstream_headers.ForwardedProto()->value().c_str() ==
+                                                 Http::Headers::get().SchemeValues.Http) {
+    return true;
+  }
+  if (connection && !connection->ssl()) {
+    return true;
+  }
+  return false;
+}
+
+bool convertRequestHeadersForInternalRedirect(Http::HeaderMap& downstream_headers,
+                                              const Http::HeaderEntry& internal_redirect,
+                                              const Network::Connection* connection) {
+  // Envoy does not currently support multiple rounds of redirects.
+  if (downstream_headers.EnvoyOriginalUrl()) {
+    return false;
+  }
+  // Make sure the redirect response contains a URL to redirect to.
+  if (internal_redirect.value().getStringView().length() == 0) {
+    return false;
+  }
+
+  Http::Utility::Url absolute_url;
+  if (!absolute_url.initialize(internal_redirect.value().getStringView())) {
+    return false;
+  }
+
+  ASSERT(connection);
+  bool scheme_is_http = schemeIsHttp(downstream_headers, connection);
+  if (scheme_is_http && absolute_url.scheme() == Http::Headers::get().SchemeValues.Https) {
+    // Don't allow serving HTTP responses over TLS.
```

Comment:
I think it's the inverse, right? "Don't allow serving HTTPS responses over plain-text" ?